README.md
Training Service
===============

The training service is a REST API used to request training asynchronously.

Overview
--------

**Start the service (development only):**
```
    python ./trainingservice.py
```

I've tested running the sensory service with uwsgi, and proxied through nginx.  Nginx configurations can be found within ./nginx of this repository.
```
uwsgi --http-socket 127.0.0.1:5000 --manage-script-name --mount /=trainingservice.py --plugin python,http --enable-threads --callable app —master
```

### Production Deployment

**Docker Container**

The recommended way to run the service is by using the official provided docker container.
The container should be deployed to a Docker Swarm as a service.

**Example**
```
docker service create \
--detach=false \
--publish 8001:80 \
--replicas 0 \
--log-driver json-file \
--name trainingservice shapeandshare/dicebox.trainingservice
```

How to apply rolling updates of the service within the swarm:
```
docker service update --image shapeandshare/dicebox.trainingservice:latest trainingservice
```

In the examples above the Docker Swarm was deployed to AWS and had the Cloudstor:aws plugin enabled and active.
The training service containers will store and read data from the shared storage.

**Global shared Cloudstor volumes mounted by all tasks in a swarm service.**

The below command is an example of how to create the shared volume within the docker swarm:
```
docker volume create -d "cloudstor:aws" --opt backing=shared dicebox
```


Contributing
------------
1. Fork the repository on Github
2. Create a named feature branch (like `add_component_x`)
3. Write your change
4. Write tests for your change (if applicable)
5. Run the tests, ensuring they all pass
6. Submit a Pull Request using Github

License and Authors
-------------------
MIT License

Copyright (c) 2017-2019 Joshua C. Burt

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

CHANGELOG.md
Change Log
==========


03.09.2019
----------
* Added Change Log!
* Updated copyright dates
* Updated requirements.txt dependencies to the latest
* Updated dicebox.core libraries to the latest
* Added addition gitignore entries
* Added code comments
* Updated 401 to 403 for auth failure
* Response for async request now uses 202 created, instead of 201.
* Route for /api/train/request is now a GET.

03.21.2019
----------
* Updated core libraries.

03.23.2019
----------
* Updated core libraries.

03.30.2019
----------
* Updated core libraries.
* Refactored to support latest library changes
* Updated docker build to account for new directory structure.

04.06.2019
----------
* Refactored for updated core library namespace updates.



app/training_service.py
#!flask/bin/python
from lib import dicebox_config as config
from lib import sensory_interface
from flask import Flask, jsonify, request, make_response, abort
from flask_cors import CORS, cross_origin
import base64
import logging
import json
from datetime import datetime
import os
import errno
import uuid
import numpy
import pika
#import logging
#import lib.dicebox_config as config
from lib.network import Network
#from datetime import datetime
import json


# Setup logging.
logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%m/%d/%Y %I:%M:%S %p',
    level=logging.DEBUG,
    filemode='w',
    filename="%s/training_service.log" % config.LOGS_DIR
)


app = Flask(__name__)
cors = CORS(app, resources={r"/api/*": {"origins": "http://localhost:*"}})


def train_request():
    training_request_id = uuid.uuid4()

    try:
        ## Submit our message
        url = config.TRAINING_SERVICE_RABBITMQ_URL
        logging.debug(url)
        parameters = pika.URLParameters(url)
        connection = pika.BlockingConnection(parameters=parameters)

        channel = connection.channel()

        channel.queue_declare(queue=config.TRAINING_SERVICE_RABBITMQ_TRAIN_REQUEST_TASK_QUEUE, durable=True)

        training_request = {}
        training_request['training_request_id'] = str(training_request_id)
        channel.basic_publish(exchange=config.TRAINING_SERVICE_RABBITMQ_EXCHANGE,
                              routing_key=config.TRAINING_SERVICE_RABBITMQ_TRAINING_REQUEST_ROUTING_KEY,
                              body=json.dumps(training_request),
                              properties=pika.BasicProperties(
                                  delivery_mode=2,  # make message persistent
                              ))
        logging.debug(" [x] Sent %r" % json.dumps(training_request))
        connection.close()
    except:
        # something went wrong..
        logging.error('we had a failure sending the request to the message system')
        return None

    return training_request_id


# for small batches..
@app.route('/api/train/request', methods=['POST'])
def make_api_train_request_public():
    if request.headers['API-ACCESS-KEY'] != config.API_ACCESS_KEY:
        logging.debug('bad access key')
        abort(401)
    if request.headers['API-VERSION'] != config.API_VERSION:
        logging.debug('bad access version')
        abort(400)
    training_request_id = train_request()
    return make_response(jsonify({'training_request_id': training_request_id}), 201)



@app.route('/api/version', methods=['GET'])
def make_api_version_public():
    return make_response(jsonify({'version':  str(config.API_VERSION)}), 201)


@app.route('/health/plain', methods=['GET'])
@cross_origin()
def make_health_plain_public():
    return make_response('true', 201)


@app.errorhandler(404)
def not_found(error):
    return make_response(jsonify({'error': 'Not found'}), 404)


if __name__ == '__main__':
    logging.debug('starting flask app')
    app.run(debug=config.FLASK_DEBUG, host=config.LISTENING_HOST, threaded=True)


legacy/lonestar_train.py
import logging
import lib.dicebox_config as config
from lib.network import Network
from datetime import datetime
import json

# Setup logging.
logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%m/%d/%Y %I:%M:%S %p',
    level=logging.DEBUG,
    filemode='w',
    filename="%s/lonestar_train.log" % config.LOGS_DIR
)


def main():
    network = Network(config.NN_PARAM_CHOICES)
    if config.LOAD_BEST_WEIGHTS_ON_START is True:
        logging.info('attempting to restart training from previous session..')
        network.create_lonestar(create_model=True,
                                weights_filename="%s/%s" % (config.WEIGHTS_DIR, config.MODEL_WEIGHTS_FILENAME))
    else:
        network.create_lonestar(create_model=True)

    with open('./category_map.json', 'w') as category_mapping_file:
        category_mapping_file.write(json.dumps(network.fsc.CATEGORY_MAP))

    i = 1
    while i <= config.EPOCHS:
        logging.info("epoch (%i of %i)" % (i, config.EPOCHS))
        network.train_and_save(config.DATASET)

        # save the model after every epoch, regardless of accuracy.
        filename = "%s/weights.epoch_%i.final.%s.hdf5" % (config.WEIGHTS_DIR, i, datetime.now().strftime('%Y-%m-%d_%H_%M_%S_%f'))
        logging.info("saving model weights after epoch %i to file %s" % (i, filename))
        network.save_model(filename)

        # the next epoch..
        i += 1

    logging.info("network accuracy: %.2f%%" % (network.accuracy * 100))
    logging.info('-'*80)
    network.print_network()

if __name__ == '__main__':
    main()


app/trainingservice.py
#!flask/bin/python
###############################################################################
# Training Service
#   Handles requests for training sessions
#
# Copyright (c) 2017-2019 Joshua Burt
###############################################################################


###############################################################################
# Dependencies
###############################################################################
from flask import Flask, jsonify, request, make_response, abort
from flask_cors import CORS, cross_origin
import logging
import uuid
import pika
import json
import os
import errno
from dicebox.config.dicebox_config import DiceboxConfig

# Config
config_file='./dicebox.config'
CONFIG = DiceboxConfig(config_file)


###############################################################################
# Allows for easy directory structure creation
# https://stackoverflow.com/questions/273192/how-can-i-create-a-directory-if-it-does-not-exist
###############################################################################
def make_sure_path_exists(path):
    try:
        if os.path.exists(path) is False:
            os.makedirs(path)
    except OSError as exception:
        if exception.errno != errno.EEXIST:
            raise


###############################################################################
# Setup logging.
###############################################################################
make_sure_path_exists(CONFIG.LOGS_DIR)
logging.basicConfig(
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%m/%d/%Y %I:%M:%S %p',
    level=logging.DEBUG,
    filemode='w',
    filename="%s/trainingservice.%s.log" % (CONFIG.LOGS_DIR, os.uname()[1])
)


###############################################################################
# Create the flask, and cors config
###############################################################################
app = Flask(__name__)
cors = CORS(app, resources={r"/api/*": {"origins": "http://localhost:*"}})


###############################################################################
# Handles submission of the actual training request to the message system
###############################################################################
def train_request():
    training_request_id = uuid.uuid4()

    try:
        ## Submit our message
        url = CONFIG.TRAINING_SERVICE_RABBITMQ_URL
        logging.debug(url)
        parameters = pika.URLParameters(url)
        connection = pika.BlockingConnection(parameters=parameters)

        channel = connection.channel()

        channel.queue_declare(queue=CONFIG.TRAINING_SERVICE_RABBITMQ_TRAIN_REQUEST_TASK_QUEUE, durable=True)

        training_request = {}
        training_request['training_request_id'] = str(training_request_id)
        channel.basic_publish(exchange=CONFIG.TRAINING_SERVICE_RABBITMQ_EXCHANGE,
                              routing_key=CONFIG.TRAINING_SERVICE_RABBITMQ_TRAINING_REQUEST_ROUTING_KEY,
                              body=json.dumps(training_request),
                              properties=pika.BasicProperties(
                                  delivery_mode=2,  # make message persistent
                              ))
        logging.debug(" [x] Sent %r" % json.dumps(training_request))
    except:
        # something went wrong..
        training_request_id = None
        logging.error('we had a failure sending the request to the message system')
    finally:
        connection.close()

    return training_request_id


###############################################################################
# Accepts requests for async training sessions
###############################################################################
@app.route('/api/train/request', methods=['GET'])
def make_api_train_request_public():
    if request.headers['API-ACCESS-KEY'] != CONFIG.API_ACCESS_KEY:
        logging.debug('bad access key')
        abort(403)
    if request.headers['API-VERSION'] != CONFIG.API_VERSION:
        logging.debug('bad access version')
        abort(400)
    training_request_id = train_request()
    return make_response(jsonify({'training_request_id': training_request_id}), 202)


###############################################################################
# Returns API version
###############################################################################
@app.route('/api/version', methods=['GET'])
def make_api_version_public():
    return make_response(jsonify({'version':  str(CONFIG.API_VERSION)}), 200)


###############################################################################
# Super generic health end-point
###############################################################################
@app.route('/health/plain', methods=['GET'])
@cross_origin()
def make_health_plain_public():
    return make_response('true', 200)


###############################################################################
# 404 Handler
###############################################################################
@app.errorhandler(404)
def not_found(error):
    return make_response(jsonify({'error': 'Not found'}), 404)


###############################################################################
# main entry point, thread safe
###############################################################################
if __name__ == '__main__':
    logging.debug('starting flask app')
    app.run(debug=CONFIG.FLASK_DEBUG, host=CONFIG.LISTENING_HOST, threaded=True)
